### PR TITLE
Allow explicit load-balancer names on AWS

### PR DESCRIPTION
**Pull request by john-dodson**
…r sorting and management, particularly in AWS

---

This change is [Reviewable](http://reviewable.k8s.io/reviews/kubernetes/kubernetes/22829)

### CODE REVIEW

**Comment by john-dodson** [comment]
I signed it!
Organization: Major League Soccer

**Comment by john-dodson** [comment]
I signed it!
Organization: Major League Soccer

**Comment by john-dodson** [comment]
@erictune My organization, Major League Soccer, has signed the CLA. Can we get a review of this? Having the load balancer name in AWS just be a GUID makes them unmanageable and, therefore, unusable. We would like to be able to provide a name prefix to Kubernetes when it creates the load balancer so that we can meaningfully sort and identify our elastic load balancers by name.

**Comment by smerrill** [comment]
This is a great improvement to manageability in an AWS environment, especially in one where k8s is being included with other legacy apps.

**Comment by bprashanth** [comment]
Ideally, the AWS implementation should just assign easy to use prefixes. If that's not possible, or this significantly improves the UX in spite of the AWS implementation's best efforts, it should get piped through as an annotation. The Service spec is not the place for it, since it has a very narrow use case. 

Here's an example of one such feature we piped through annotations: https://github.com/kubernetes/kubernetes/pull/21431

@justinsb @therc for thoughts on AWS

**Comment by thockin** [comment]
Yeah, we're not adding this to the API.  I'd be happy to take a change that reads an annotation and uses that in the AWS code.  Or that simply made the AWS code use a better name...

**Comment by thockin** [comment]
Sorry.  I appreciate the pull request and the work that went into it.  I'll levae this open because I think youc an turn it around pretty easily into something workable.  Feel free to close it if you would prefer to do a new PR instead.

**Comment by justinsb** [comment]
The challenge with AWS is that the length is limited (32 characters I believe) and so you probably have to deal with collisions if you allow long prefixes.  Maybe we could base-32 encode the UUID (instead of base-16), and then we could argue that we can make do with e.g. 12 id characters leaving the rest for a prefix.

However, we did add additional tags with the service name and other details in 1.2.  Does that not help with the management/identification?  Is there a particular tag you'd like?

**Comment by thockin** [comment]
Yeah, we started using description fields in GCE..

**Comment by john-dodson** [comment]
It doesn't help because the AWS console section for ELBs doesn't support tag filtering for some reason.

**Comment by john-dodson** [comment]
@thockin I updated the PR to use an annotation instead of an API change. Please review.

File: pkg/cloudprovider/providers/aws/aws.go [hunk 1]
```diff
@@ -68,6 +68,7 @@ const TagNameSubnetPublicELB = "kubernetes.io/role/elb"
 // Currently we accept only the value "0.0.0.0/0" - other values are an error.
 // This lets us define more advanced semantics in future.
 const ServiceAnnotationLoadBalancerInternal = "service.beta.kubernetes.io/aws-load-balancer-internal"
+const ServiceAnnotationLoadBalancerName = "service.beta.kubernetes.io/aws-load-balancer-name"
```
Comment (thockin) [inline pkg/cloudprovider/providers/aws/aws.go]: needs a comment, especially that this is a full name, and not a prefix.

@justinsb are you ok with a full name here?

**Comment by thockin** [comment]
Change looks OK, needs comments.  We need to sort out your CLA

**Comment by thockin** [comment]
Re-titleing for relnote

**Comment by thockin** [comment]
I think your github account needs to have an email that is a member of a mailing list that is associated with the CLA.  I can confirm MLS is in the signee list, but I can't confirm you work there :)

File: pkg/cloudprovider/providers/aws/aws.go (hunk 1 above)
Comment (thockin) [inline pkg/cloudprovider/providers/aws/aws.go]: We also need a plan to move these out of "beta" and into "real".  @justinsb do you feel like you need beta for this?

`aws.service.kubernetes.io/load-balancer-name` or `service.aws.kubernetes.io/load-balancer-name` ?

**Comment by john-dodson** [comment]
@thockin I replied to your email and I added my MLS email to my GitHub account.

**Comment by john-dodson** [comment]
@thockin Can you mark this as OK to test?

**Comment by therc** [comment]
With the current base-16 encoding, we are preserving only 124 bits of the service's UUID. It's not 128, because GCE requires the name to start with a lowercase letter, so they all have a leading `a`. With base-32 we would go up to 155 bits (31 * 5). With a little more work and trickier math (base-36), we'd gain another 5.2 bits.

Maybe we could allow a prefix of up to sixteen characters (dash included), which would give us a minimum of 64/80/82.7 bits.

**Comment by thockin** [comment]
Adding Justin.  I feel bad that this one fell through the cracks

**Comment by thockin** [comment]
ok to test

**Comment by thockin** [comment]
ping @justinsb what to do with this one?

File: pkg/cloudprovider/providers/aws/aws.go (hunk 1 above)
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws.go]: I would prefer to have the alpha/beta status in the external docs.

Is there some reason you put aws into the prefix @thockin?

My preference would be `service.kubernetes.io/load-balancer-name` or `service.kubernetes.io/aws-load-balancer-name`.  The former if we think think might be useful on multiple clouds.

File: pkg/cloudprovider/providers/aws/aws.go [hunk 2]
```diff
@@ -2159,7 +2160,14 @@ func (s *AWSCloud) EnsureLoadBalancer(apiService *api.Service, hosts []string, a
 		return nil, fmt.Errorf("could not find any suitable subnets for creating the ELB")
 	}
 
-	loadBalancerName := cloudprovider.GetLoadBalancerName(apiService)
+	loadBalancerName := annotations[ServiceAnnotationLoadBalancerName]
+	if loadBalancerName != "" {
+		if len(loadBalancerName) > 32 {
+			return nil, fmt.Errorf("annotation %q=%q detected, but AWS only allows load balancer names up to 32 bytes", ServiceAnnotationLoadBalancerName, loadBalancerName)
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws.go]: We could also truncate, but I guess this is better.

If we intend to allow a cross-cloud name annotation, we should probably truncate & perform any obvious sanitization required (although the AWS API won't let us create an ELB with an illegal name anyway)

**Comment by justinsb** [comment]
I'm also sorry for not reviewing this before!

So actually this LGTM, though I think @thockin we need your input to resolve the name of the annotation.

If we enter the name of an existing ELB, it will be reused now.  That's good functionality that users have been asking for, but I was thinking we would do this using a shared secret approach (e.g. https://github.com/kubernetes/kubernetes/issues/23483).

Security issues for you to OK @thockin : 
* I think that e.g. if we point two clusters or even two services at the same ELB "bad things" will probably happen.
* I think this allows "takeover" of any ELB in the AWS account - i.e. we break cluster isolation.

**Comment by thockin** [comment]
Here's how I see it.  Ideally, each platform chooses the names that make the best UX for users of that platform.  If your platform has a severe limitation (I don't think 32 is THAT severe, but let's run with it), then it's on you to find a way to manage it.  I'm not going to add this as a generic field for all platforms (I think), which is what a universal annotation approximates.  I'd rather see the platform do something sane by default.  GCE has a very similar problem - while we populate metadata, we can't see that in most UIs.  I just have a screen full of hashed IDs.  We should fix it (but compat is hell and nobody has bothered to work out the solution yet).

If you want to do a prefix that users can provide, I won't stop you but I don't think it is generic.  Unless I am wrong and everyone glows about the UX, which means I am 180 degrees off, which would not surprise me :)

**Comment by justinsb** [comment]
@john-dodson  If this is "just" about manageability in the API, how about we make the label a name _hint_ - not a full name.

If you cloud allows arbitrary labels, great - we use the name you specify.  But otherwise we try to use it as a prefix or similar, in a way that is cloud specific.

Likely on AWS or GCE, we would use the first N characters of the label, with invalid characters dropped, such that we still had enough characters remaining to make collisions vanishingly unlikely.  (We already only use 124 of the 128 bits or so).

Another wrinkle: if the user changes this value, we will leak ELBs.  I think the solution there would be to make sure that ELBs have the service ID in their labels (I believe they do), and then to remove any ELBs for the service that don't match.  We can also use this to make sure there is no collision, by verifying the service ids match.

At this stage in the cycle, my gut feeling is that it is too risky.  I know that's my fault for not reviewing this sooner, and I can only apologize.  If you open an issue, we can debate the various options and think through the consequences, and hopefully get this into 1.5 which is still scheduled for this year I believe.

**Comment by thockin** [comment]
If it is a hint, just use the service namespace/name?  Ugh, now we're back
to where we were. Leaking is a real concern.

**Comment by alexw23** [comment]
Can this be merged?

**Comment by jeremyd** [comment]
This would have been really great to have +1  sorry it didn't work out.

**Pull request closed without merging**